File: src/ReverseLinkedList.java
// T: O(n)
// S: O(1)

public class ReverseLinkedList {

    public static ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }

        ListNode current = head, ahead = head.next, temp = head.next.next;
        head.next = null;

        while (temp != null) {
            ahead.next = current;
            current = ahead;
            ahead = temp;
            temp = temp.next;
        }
        ahead.next = current;
        return ahead;
    }
}

File: python/reverse_linked_list.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def reverseList(self, head: ListNode) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        a, b, c = head, head.next, head.next.next
        a.next = None
        while c is not None:
            b.next = a
            a, b, c = b, c, c.next
        b.next = a
        return b
